Adding more features in second iteration. Following are the new approaches tried:
<br>
- Imputing the missing values in train and test data.
<br>
- Extracting features from the two transaction files separately
- Mean Encoding the categorical variables instead of One hot encoding.
- Generating additional features based on existing features.

# Google Drive setup 

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
cd /content/drive/My Drive/Applied\ AI\ course\ Assignments/Case\ Study\ 1

/content/drive/My Drive/Applied AI course Assignments/Case Study 1


In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import datetime

from statsmodels.stats.outliers_influence import variance_inflation_factor

import gc
import warnings
warnings.filterwarnings('ignore')

#plt.style.use('dark_background')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


# Script to reduce memory usage

In [ ]:
#https://www.kaggle.com/fabiendaniel/elo-world
#Function to load data into pandas and reduce memory usage

def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage().sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage().sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

# Train and test csv files

In [ ]:
train = pd.read_csv('Data/train.csv')
test = pd.read_csv('Data/test.csv')

In [ ]:
print(train.info())
print(test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201917 entries, 0 to 201916
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   first_active_month  201917 non-null  object 
 1   card_id             201917 non-null  object 
 2   feature_1           201917 non-null  int64  
 3   feature_2           201917 non-null  int64  
 4   feature_3           201917 non-null  int64  
 5   target              201917 non-null  float64
dtypes: float64(1), int64(3), object(2)
memory usage: 9.2+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123623 entries, 0 to 123622
Data columns (total 5 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   first_active_month  123622 non-null  object
 1   card_id             123623 non-null  object
 2   feature_1           123623 non-null  int64 
 3   feature_2           123623 non-null  int64 
 4   feature_3     

So our train data has 201917 cards.
<br> 
We have to make sure we generate features involving these cards only to avoid data leakage.

Referring to Feature Engineering by: https://www.kaggle.com/mfjwr1/simple-lightgbm-without-blending

In [ ]:
train.isna().any()

first_active_month    False
card_id               False
feature_1             False
feature_2             False
feature_3             False
target                False
dtype: bool

In [ ]:
test.isna().any()

first_active_month     True
card_id               False
feature_1             False
feature_2             False
feature_3             False
dtype: bool

In [ ]:
test[test['first_active_month'].isna()]

first_active_month          card_id  feature_1  feature_2  feature_3
11578                NaN  C_ID_c27b4f80f7          5          2          1

In [ ]:
test['first_active_month'].fillna(test['first_active_month'].mode()[0], inplace=True)

In [ ]:
test.isna().any()

first_active_month    False
card_id               False
feature_1             False
feature_2             False
feature_3             False
dtype: bool

#### Features using first_active_month

In [ ]:
train['first_active_month'] = pd.to_datetime(train['first_active_month'])
train['quarter_first_active_month'] = train['first_active_month'].dt.quarter
train['first_active_month_diff_from_today'] = (datetime.datetime.today() - train['first_active_month']).dt.days

test['first_active_month'] = pd.to_datetime(test['first_active_month'])
test['quarter_first_active_month'] = test['first_active_month'].dt.quarter
test['first_active_month_diff_from_today'] = (datetime.datetime.today() - test['first_active_month']).dt.days

In [ ]:
train.head()

first_active_month  ... first_active_month_diff_from_today
0         2017-06-01  ...                               1215
1         2017-01-01  ...                               1366
2         2016-08-01  ...                               1519
3         2017-09-01  ...                               1123
4         2017-11-01  ...                               1062

[5 rows x 8 columns]

In [ ]:
test.head()

first_active_month  ... first_active_month_diff_from_today
0         2017-04-01  ...                               1276
1         2017-01-01  ...                               1366
2         2017-08-01  ...                               1154
3         2017-12-01  ...                               1032
4         2015-12-01  ...                               1763

[5 rows x 7 columns]

### Labeling rare data points 

Using mean encoding for categorical features points to predict rare data points better

In [ ]:
train['rare_datapoints'] = 0
train.loc[train['target'] < -30, 'rare_datapoints'] = 1

In [ ]:
train['rare_datapoints'].value_counts()

0    199710
1      2207
Name: rare_datapoints, dtype: int64

In [ ]:
for f in ['feature_1', 'feature_2', 'feature_3']:
  rare_data_mean = train.groupby([f])['rare_datapoints'].mean()
  train[f] = train[f].map(rare_data_mean)
  test[f] = test[f].map(rare_data_mean)

In [ ]:
train.head()

first_active_month  ... rare_datapoints
0         2017-06-01  ...               0
1         2017-01-01  ...               0
2         2016-08-01  ...               0
3         2017-09-01  ...               0
4         2017-11-01  ...               0

[5 rows x 9 columns]

In [ ]:
test.head()

first_active_month  ... first_active_month_diff_from_today
0         2017-04-01  ...                               1276
1         2017-01-01  ...                               1366
2         2017-08-01  ...                               1154
3         2017-12-01  ...                               1032
4         2015-12-01  ...                               1763

[5 rows x 7 columns]

#### Combining categorical features and date features

In [ ]:
test['target'] = np.nan
#Products
train['cat_time_feature1'] = train['first_active_month_diff_from_today'] * train['feature_1']
train['cat_time_feature2'] = train['first_active_month_diff_from_today'] * train['feature_2']
train['cat_time_feature3'] = train['first_active_month_diff_from_today'] * train['feature_3']

test['cat_time_feature1'] = test['first_active_month_diff_from_today'] * test['feature_1']
test['cat_time_feature2'] = test['first_active_month_diff_from_today'] * test['feature_2']
test['cat_time_feature3'] = test['first_active_month_diff_from_today'] * test['feature_3']

#Ratios
train['cat_time_ratio1'] = train['feature_1'] / train['first_active_month_diff_from_today']
train['cat_time_ratio2'] = train['feature_2'] / train['first_active_month_diff_from_today']
train['cat_time_ratio3'] = train['feature_3'] / train['first_active_month_diff_from_today']

test['cat_time_ratio1'] = test['feature_1'] / test['first_active_month_diff_from_today']
test['cat_time_ratio2'] = test['feature_2'] / test['first_active_month_diff_from_today']
test['cat_time_ratio3'] = test['feature_3'] / test['first_active_month_diff_from_today']

In [ ]:
train.head()

first_active_month          card_id  ...  cat_time_ratio2  cat_time_ratio3
0         2017-06-01  C_ID_92a2005557  ...         0.000007         0.000009
1         2017-01-01  C_ID_3d0044924f  ...         0.000008         0.000008
2         2016-08-01  C_ID_d639edf6cd  ...         0.000006         0.000007
3         2017-09-01  C_ID_186d6a6901  ...         0.000013         0.000009
4         2017-11-01  C_ID_cdbd2c0db2  ...         0.000013         0.000010

[5 rows x 15 columns]

### Aggregation based features on categorical features

In [ ]:
train['feature_sum'] = train['feature_1'] + train['feature_2'] + train['feature_3']
train['feature_mean'] = train['feature_sum']/3
train['feature_max'] = train[['feature_1', 'feature_2', 'feature_3']].max(axis=1)
train['feature_min'] = train[['feature_1', 'feature_2', 'feature_3']].min(axis=1)
train['feature_std'] = train[['feature_1', 'feature_2', 'feature_3']].std(axis=1)

feature_cols = ['feature_1', 'feature_2', 'feature_3']
for f in feature_cols:
    train['days_' + f] = train['first_active_month_diff_from_today'] * train[f]
    train['days_' + f + '_ratio'] = train[f] / train['first_active_month_diff_from_today']

test['feature_sum'] = test['feature_1'] + test['feature_2'] + test['feature_3']
test['feature_mean'] = test['feature_sum']/3
test['feature_max'] = test[['feature_1', 'feature_2', 'feature_3']].max(axis=1)
test['feature_min'] = test[['feature_1', 'feature_2', 'feature_3']].min(axis=1)
test['feature_std'] = test[['feature_1', 'feature_2', 'feature_3']].std(axis=1)

feature_cols = ['feature_1', 'feature_2', 'feature_3']
for f in feature_cols:
    test['days_' + f] = test['first_active_month_diff_from_today'] * test[f]
    test['days_' + f + '_ratio'] = test[f] / test['first_active_month_diff_from_today']

In [ ]:
train.head()

first_active_month          card_id  ...  days_feature_3  days_feature_3_ratio
0         2017-06-01  C_ID_92a2005557  ...       13.884438              0.000009
1         2017-01-01  C_ID_3d0044924f  ...       14.046353              0.000008
2         2016-08-01  C_ID_d639edf6cd  ...       15.619626              0.000007
3         2017-09-01  C_ID_186d6a6901  ...       11.547624              0.000009
4         2017-11-01  C_ID_cdbd2c0db2  ...       10.920371              0.000010

[5 rows x 26 columns]

In [ ]:
test.head()

first_active_month          card_id  ...  days_feature_3  days_feature_3_ratio
0         2017-04-01  C_ID_0ab67a22ab  ...       14.581516              0.000009
1         2017-01-01  C_ID_130fd0cbdd  ...       14.046353              0.000008
2         2017-08-01  C_ID_b709037bc5  ...       13.187359              0.000010
3         2017-12-01  C_ID_d27d835a9f  ...       10.611886              0.000010
4         2015-12-01  C_ID_2b5e3df5c2  ...       20.146719              0.000006

[5 rows x 25 columns]

In [ ]:
train.isna().any().sum()

0

In [ ]:
test.isna().any().sum() #Checking except for target which is nan

1

In [ ]:
train.to_csv('New_Data_processed/processed_train.csv')
test.to_csv('New_Data_processed/processed_test.csv')

# Transaction Files

## 1. historical_transactions 

In [ ]:
historical_transactions = reduce_mem_usage(pd.read_csv('Data/historical_transactions.csv'))
historical_transactions.replace([-np.inf, np.inf], np.nan, inplace=True)

Mem. usage decreased to 1749.11 Mb (43.7% reduction)


In [ ]:
historical_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29112361 entries, 0 to 29112360
Data columns (total 14 columns):
 #   Column                Dtype  
---  ------                -----  
 0   authorized_flag       object 
 1   card_id               object 
 2   city_id               int16  
 3   category_1            object 
 4   installments          int16  
 5   category_3            object 
 6   merchant_category_id  int16  
 7   merchant_id           object 
 8   month_lag             int8   
 9   purchase_amount       float32
 10  purchase_date         object 
 11  category_2            float16
 12  state_id              int8   
 13  subsector_id          int8   
dtypes: float16(1), float32(1), int16(3), int8(3), object(6)
memory usage: 1.7+ GB


In [ ]:
#all_transactions['installments'].replace(-1, np.nan, inplace=True)
#all_transactions['installments'].replace(999, np.nan, inplace=True)
#all_transactions['installments'].fillna(all_transactions['installments'].mode()[0], inplace=True)

In [ ]:
historical_transactions.isna().any()

authorized_flag         False
card_id                 False
city_id                 False
category_1              False
installments            False
category_3               True
merchant_category_id    False
merchant_id              True
month_lag               False
purchase_amount         False
purchase_date           False
category_2               True
state_id                False
subsector_id            False
dtype: bool

Imputing the missing values

In [ ]:
historical_transactions['merchant_id'].fillna(historical_transactions['merchant_id'].mode()[0], inplace=True)
historical_transactions['category_3'].fillna(historical_transactions['category_3'].mode()[0], inplace=True)
historical_transactions['category_2'].fillna(historical_transactions['category_2'].mode()[0], inplace=True)

In [ ]:
historical_transactions['purchase_amount'] = historical_transactions['purchase_amount'].apply(lambda x: min(x, 0.8))
historical_transactions['installments'].replace([-1, 999], np.nan, inplace=True)
historical_transactions['installments'].fillna(historical_transactions['installments'].mode()[0], inplace=True)
historical_transactions['price'] = historical_transactions['purchase_amount'] / (historical_transactions['installments'] + 0.01)

In [ ]:
historical_transactions.isna().any()

authorized_flag         False
card_id                 False
city_id                 False
category_1              False
installments            False
category_3              False
merchant_category_id    False
merchant_id             False
month_lag               False
purchase_amount         False
purchase_date           False
category_2              False
state_id                False
subsector_id            False
price                   False
dtype: bool

In [ ]:
historical_transactions['authorized_flag'] = historical_transactions['authorized_flag'].map({'Y': 1, 'N': 0})
historical_transactions['category_1'] = historical_transactions['category_1'].map({'Y': 1, 'N': 0})

In [ ]:
historical_transactions.head()

authorized_flag          card_id  city_id  ...  state_id  subsector_id      price
0                1  C_ID_4e6213e9bc       88  ...        16            37 -70.333093
1                1  C_ID_4e6213e9bc       88  ...        16            16 -73.312849
2                1  C_ID_4e6213e9bc       88  ...        16            37 -72.038603
3                1  C_ID_4e6213e9bc       88  ...        16            34 -73.535240
4                1  C_ID_4e6213e9bc       88  ...        16            37 -72.286540

[5 rows x 15 columns]

In [ ]:
historical_transactions.to_csv('New_Data_processed/historical_transactions_processed.csv')

### Featurization using purchase date

Referring the EDA kernel on dates: https://www.kaggle.com/denzo123/a-closer-look-at-date-variables

In [ ]:
historical_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/historical_transactions_processed.csv', index_col=0))

Mem. usage decreased to 1582.53 Mb (52.5% reduction)


In [ ]:
historical_transactions = historical_transactions[['card_id', 'purchase_date', 'month_lag']]

In [ ]:
pd.to_datetime(historical_transactions['purchase_date'], format='%Y-%m-%d %H:%M:%S').dt.year.value_counts()

2017    23828069
2018     5284292
Name: purchase_date, dtype: int64

So our data contains transactions from two years: 2017 and 2018.

In [ ]:
historical_transactions['purchase_date'] = pd.to_datetime(historical_transactions['purchase_date'], format='%Y-%m-%d %H:%M:%S')

In [ ]:
# Get the day of the week for the purchase date
def get_weekday(data):
  return data.dt.dayofweek

# Return 1 if the purchase date is on a weekend
def is_weekend(day):
  if day == 5 or day == 6:
    return 1
  else:
    return 0

#Return day of purchase
def get_day(date_obj):
  return date_obj.dt.day

#Return week of the year of purchase
def get_week_of_year(date_obj):
  return date_obj.dt.weekofyear

#Return hour of purchase
def get_hour(date_obj):
  return date_obj.dt.hour

#Return month of purchase date
def get_purchase_month(data):
  return data.dt.month

#Return the month phase during the purchase
#Eg. Early in the month, mid-month or at the end of the month
def get_time_of_month(date):
  if date.day <=10:
    return "Early"
  elif date.day > 10 and date.day <= 20:
    return "Mid"
  else:
    return "End"
  

# Time of the day during purchase
#Eg. Morning, Afternoon, Evening, Night
def get_time_of_day(time):
  if time.hour >= 4 and time.hour < 12:
    return "Morning"
  elif time.hour >= 12 and time.hour < 17:
    return "Afternoon"
  elif time.hour >= 17 and time.hour < 22:
    return "Evening"
  else:
    return "Night"
    

# Returns 1 if the purchase was made on a holiday(Saturdays and sundays excluded)
# Google Search : list of holidays in brazil 2017 and 2018
def get_isholiday(date):
  holiday_list=[
            '01-01-17', '14-02-17', '28-08-17', '14-04-17', '16-04-17', '21-04-17',
            '01-05-17', '15-06-17', '07-09-17', '12-10-17', '02-11-17', '15-11-17', 
            '24-12-17', '25-12-17', '31-12-17',
            '01-01-18', '14-02-18', '28-08-18', '14-04-18', '16-04-18', '21-04-18',
            '01-05-18', '15-06-18', '07-09-18', '12-10-18', '02-11-18', '15-11-18', 
            '24-12-18', '25-12-18', '31-12-18'
  ]
  date = date.strftime(format='%d-%m-%y') 
  if date in holiday_list:
    return 1
  else:
    return 0

In [ ]:
historical_transactions.head()

card_id       purchase_date  month_lag
0  C_ID_4e6213e9bc 2017-06-25 15:33:07         -8
1  C_ID_4e6213e9bc 2017-07-15 12:10:45         -7
2  C_ID_4e6213e9bc 2017-08-09 22:04:29         -6
3  C_ID_4e6213e9bc 2017-09-02 10:06:26         -5
4  C_ID_4e6213e9bc 2017-03-10 01:14:19        -11

In [ ]:
historical_transactions['weekday'] = get_weekday(historical_transactions['purchase_date'])
historical_transactions['is_weekend'] = historical_transactions['weekday'].apply(lambda day: is_weekend(day))
historical_transactions['purchase_month'] = get_purchase_month(historical_transactions['purchase_date'])
historical_transactions['purchase_day'] = get_day(historical_transactions['purchase_date'])
historical_transactions['week_of_year'] = get_week_of_year(historical_transactions['purchase_date'])
historical_transactions['purchase_hour'] = get_hour(historical_transactions['purchase_date'])
#historical_transactions['purchase_time_of_day'] = historical_transactions['purchase_date'].apply(lambda day: get_time_of_day(day))
#historical_transactions['purchase_month_time'] = historical_transactions['purchase_date'].apply(lambda date: get_time_of_month(date))
historical_transactions['purchase_on_holiday'] = historical_transactions['purchase_date'].apply(lambda date_obj: get_isholiday(date_obj))
historical_transactions['purchase_date'] = historical_transactions['purchase_date'].dt.date

In [ ]:
historical_transactions['month_diff'] = ((datetime.date.today() - historical_transactions['purchase_date']).dt.days)//30
historical_transactions['month_diff'] +=historical_transactions['month_lag']
del historical_transactions['month_lag']

Saving the dataframe due to memory constraints.

In [ ]:
historical_transactions.to_csv('New_Data_processed/historical_transactions_date_features.csv')

#### Aggregating:

In [ ]:
historical_transactions = pd.read_csv('New_Data_processed/historical_transactions_date_features.csv', index_col=0)

In [ ]:
historical_transactions.head()

In [ ]:
historical_transactions['purchase_date'] = pd.to_datetime(historical_transactions['purchase_date'])

In [ ]:
aggregations = {
    'is_weekend': ['sum', 'mean'],
    'purchase_on_holiday': ['sum', 'mean'],
    'weekday' : ['nunique', 'sum', 'mean'],
    'purchase_hour': ['nunique', 'mean', 'min', 'max'],
    'week_of_year': ['nunique', 'mean', 'min', 'max'],
    'month_diff': ['sum', 'mean', 'min', 'max', 'var', 'skew'],
    'purchase_day': ['nunique', 'sum', 'min'],
    'purchase_date' : [np.ptp, 'min', 'max'],
    'purchase_month' : ['sum', 'mean', 'nunique']

}

aggregated_date_features = historical_transactions.groupby('card_id').agg(aggregations)
aggregated_date_features.columns = ['transactions_'+'_'.join(col).strip() 
                           for col in aggregated_date_features.columns.values]

In [ ]:
aggregated_date_features.head()

transactions_is_weekend_sum  ...  transactions_purchase_month_nunique
card_id                                       ...                                     
C_ID_00007093c1                           25  ...                                   12
C_ID_0001238066                           52  ...                                    6
C_ID_0001506ef0                           32  ...                                   11
C_ID_0001793786                           37  ...                                   10
C_ID_000183fdda                           33  ...                                    7

[5 rows x 30 columns]

In [ ]:
aggregated_date_features['transactions_purchase_date_ptp'] = aggregated_date_features['transactions_purchase_date_ptp'].dt.days

In [ ]:
d_now = datetime.datetime.today()
d_now

datetime.datetime(2020, 9, 29, 8, 19, 22, 855768)

In [ ]:
aggregated_date_features['transactions_purchase_date_max_diff_now'] = (datetime.datetime.today() - aggregated_date_features['transactions_purchase_date_max']).dt.days

In [ ]:
aggregated_date_features['transactions_purchase_date_min_diff_now'] = (datetime.datetime.today() - aggregated_date_features['transactions_purchase_date_min']).dt.days

In [ ]:
del aggregated_date_features['transactions_purchase_date_ptp']

In [ ]:
aggregated_date_features.head()

transactions_is_weekend_sum  ...  transactions_purchase_date_min_diff_now
card_id                                       ...                                         
C_ID_00007093c1                           25  ...                                     1323
C_ID_0001238066                           52  ...                                     1097
C_ID_0001506ef0                           32  ...                                     1354
C_ID_0001793786                           37  ...                                     1347
C_ID_000183fdda                           33  ...                                     1149

[5 rows x 31 columns]

In [ ]:
aggregated_date_features.to_csv('New_Data_processed/historical_transactions_date_features_final.csv')

### Other Features from transactions

In [ ]:
historical_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/historical_transactions_processed.csv', index_col = 0))

Mem. usage decreased to 1582.53 Mb (55.5% reduction)


In [ ]:
date_features = reduce_mem_usage(pd.read_csv('New_Data_processed/historical_transactions_date_features.csv', index_col = 0))

Mem. usage decreased to 888.44 Mb (63.6% reduction)


In [ ]:
month_diff = date_features['month_diff']

In [ ]:
historical_transactions['duration'] = historical_transactions['purchase_amount']*month_diff
historical_transactions['amount_month_ratio'] = historical_transactions['purchase_amount']/month_diff

In [ ]:
del month_diff,date_features

gc.collect()

0

In [ ]:
historical_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_4e6213e9bc  ... -21.803259           -0.022688
1                1  C_ID_4e6213e9bc  ... -23.460112           -0.022910
2                1  C_ID_4e6213e9bc  ... -23.052353           -0.022512
3                1  C_ID_4e6213e9bc  ... -23.531277           -0.022980
4                1  C_ID_4e6213e9bc  ... -23.131693           -0.022590

[5 rows x 17 columns]

In [ ]:
cat_aggregations = pd.DataFrame()
for col in ['category_2', 'category_3']:
  cat_aggregations[col + '_mean'] = historical_transactions.groupby([col])['purchase_amount'].transform('mean')
  cat_aggregations[col + '_min'] = historical_transactions.groupby([col])['purchase_amount'].transform('min')
  cat_aggregations[col + '_max'] = historical_transactions.groupby([col])['purchase_amount'].transform('max')
  cat_aggregations[col + '_sum'] = historical_transactions.groupby([col])['purchase_amount'].transform('sum')

In [ ]:
cat_aggregations['card_id'] = historical_transactions['card_id']

In [ ]:
cat_aggregations.head()

category_2_mean  category_2_min  ...  category_3_sum          card_id
0        -0.606656       -0.746908  ...   -9.971130e+06  C_ID_4e6213e9bc
1        -0.606656       -0.746908  ...   -9.971130e+06  C_ID_4e6213e9bc
2        -0.606656       -0.746908  ...   -9.971130e+06  C_ID_4e6213e9bc
3        -0.606656       -0.746908  ...   -9.971130e+06  C_ID_4e6213e9bc
4        -0.606656       -0.746908  ...   -9.971130e+06  C_ID_4e6213e9bc

[5 rows x 9 columns]

In [ ]:
historical_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_4e6213e9bc  ... -21.803259           -0.022688
1                1  C_ID_4e6213e9bc  ... -23.460112           -0.022910
2                1  C_ID_4e6213e9bc  ... -23.052353           -0.022512
3                1  C_ID_4e6213e9bc  ... -23.531277           -0.022980
4                1  C_ID_4e6213e9bc  ... -23.131693           -0.022590

[5 rows x 17 columns]

In [ ]:
cat_aggregations.to_csv('New_Data_processed/transactions_category_aggregations_features.csv')

In [ ]:
del cat_aggregations
gc.collect()

24

In [ ]:
historical_transactions.drop(columns=['purchase_date'], axis=1, inplace=True)

In [ ]:
historical_transactions.to_csv('New_Data_processed/historical_transactions_processed.csv')

Saving the dataframe due to memory constraints

#### Aggregating the category features generated using transform

In [ ]:
cat_aggregations = reduce_mem_usage(pd.read_csv('New_Data_processed/transactions_category_aggregations_features.csv'))

Mem. usage decreased to 888.44 Mb (60.0% reduction)


In [ ]:
aggregated = cat_aggregations.groupby('card_id').mean()

In [ ]:
aggregated.head()

Unnamed: 0  category_2_mean  ...  category_3_max  category_3_sum
card_id                                       ...                                
C_ID_00007093c1  19095848.0        -0.593750  ...        0.799805      -6147802.5
C_ID_0001238066  26539170.0        -0.605957  ...        0.799805      -5528821.5
C_ID_0001506ef0  25106432.5        -0.591309  ...        0.799805      -9930074.0
C_ID_0001793786  12034801.5        -0.601562  ...        0.799805      -9908405.0
C_ID_000183fdda   6867570.5        -0.592285  ...        0.799805      -5560250.0

[5 rows x 9 columns]

In [ ]:
aggregated.drop(columns=['Unnamed: 0'], axis=1, inplace=True)

In [ ]:
aggregated.to_csv('New_Data_processed/transactions_category_aggregated_final.csv')

#### Numerical Features aggregation

In [ ]:
historical_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/historical_transactions_processed.csv', index_col=0))

Mem. usage decreased to 1471.48 Mb (61.0% reduction)


In [ ]:
historical_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_4e6213e9bc  ... -21.796875           -0.022690
1                1  C_ID_4e6213e9bc  ... -23.453125           -0.022903
2                1  C_ID_4e6213e9bc  ... -23.046875           -0.022507
3                1  C_ID_4e6213e9bc  ... -23.531250           -0.022980
4                1  C_ID_4e6213e9bc  ... -23.125000           -0.022583

[5 rows x 16 columns]

#### Aggregating the columns based on card_id

In [ ]:
aggregations = {
    'authorized_flag' : ['sum', 'mean'],
    'category_1' : ['sum', 'mean'],
    'card_id': ['size'],
    'city_id' : ['nunique'],
    'state_id' : ['nunique'],
    'subsector_id' : ['nunique'],
    'merchant_category_id' : ['nunique'],
    'merchant_id': ['nunique'],
    'month_lag' : ['sum', 'mean', 'min', 'max', 'var'],
    'duration': ['mean', 'min', 'mean', 'max', 'var', 'skew'],
    'amount_month_ratio': ['mean', 'min', 'max', 'var', 'skew'],
    'installments' : ['sum', 'mean', 'min', 'max', 'var'],
    'purchase_amount' : ['sum', 'mean', 'min', 'max', 'var'],
    'price': ['sum', 'mean', 'min', 'max', 'var', 'skew']
    
}

aggregated_numerical_features = historical_transactions.groupby('card_id').agg(aggregations)
aggregated_numerical_features.columns = ['transactions_'+'_'.join(col).strip() 
                           for col in aggregated_numerical_features.columns.values]

In [ ]:
aggregated_numerical_features.head()

transactions_authorized_flag_sum  ...  transactions_price_skew
card_id                                            ...                         
C_ID_00007093c1                             114.0  ...                 1.458984
C_ID_0001238066                             120.0  ...                     -inf
C_ID_0001506ef0                              62.0  ...                      inf
C_ID_0001793786                             189.0  ...                      NaN
C_ID_000183fdda                             137.0  ...                     -inf

[5 rows x 42 columns]

In [ ]:
aggregated_numerical_features.to_csv('New_Data_processed/transactions_other_features_final.csv')

## 2. new_merchant_transactions

In [ ]:
new_merchant_transactions = reduce_mem_usage(pd.read_csv('Data/new_merchant_transactions.csv'))
new_merchant_transactions.replace([-np.inf, np.inf], np.nan, inplace=True)

Mem. usage decreased to 114.20 Mb (45.5% reduction)


In [ ]:
new_merchant_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1963031 entries, 0 to 1963030
Data columns (total 14 columns):
 #   Column                Dtype  
---  ------                -----  
 0   authorized_flag       object 
 1   card_id               object 
 2   city_id               int16  
 3   category_1            object 
 4   installments          int16  
 5   category_3            object 
 6   merchant_category_id  int16  
 7   merchant_id           object 
 8   month_lag             int8   
 9   purchase_amount       float16
 10  purchase_date         object 
 11  category_2            float16
 12  state_id              int8   
 13  subsector_id          int8   
dtypes: float16(2), int16(3), int8(3), object(6)
memory usage: 114.2+ MB


In [ ]:
new_merchant_transactions.isna().any()

authorized_flag         False
card_id                 False
city_id                 False
category_1              False
installments            False
category_3               True
merchant_category_id    False
merchant_id              True
month_lag               False
purchase_amount         False
purchase_date           False
category_2               True
state_id                False
subsector_id            False
dtype: bool

Imputing the missing values

In [ ]:
new_merchant_transactions['merchant_id'].fillna(new_merchant_transactions['merchant_id'].mode()[0], inplace=True)
new_merchant_transactions['category_3'].fillna(new_merchant_transactions['category_3'].mode()[0], inplace=True)
new_merchant_transactions['category_2'].fillna(new_merchant_transactions['category_2'].mode()[0], inplace=True)

In [ ]:
new_merchant_transactions['purchase_amount'] = new_merchant_transactions['purchase_amount'].apply(lambda x: min(x, 0.8))
new_merchant_transactions['installments'].replace([-1, 999], np.nan, inplace=True)
new_merchant_transactions['installments'].fillna(new_merchant_transactions['installments'].mode()[0], inplace=True)
new_merchant_transactions['price'] = new_merchant_transactions['purchase_amount'] / (new_merchant_transactions['installments'] + 0.01)

In [ ]:
new_merchant_transactions.isna().any()

authorized_flag         False
card_id                 False
city_id                 False
category_1              False
installments            False
category_3              False
merchant_category_id    False
merchant_id             False
month_lag               False
purchase_amount         False
purchase_date           False
category_2              False
state_id                False
subsector_id            False
price                   False
dtype: bool

In [ ]:
new_merchant_transactions['authorized_flag'] = new_merchant_transactions['authorized_flag'].map({'Y': 1, 'N': 0})
new_merchant_transactions['category_1'] = new_merchant_transactions['category_1'].map({'Y': 1, 'N': 0})

In [ ]:
new_merchant_transactions.head()

authorized_flag          card_id  city_id  ...  state_id  subsector_id     price
0                1  C_ID_415bb3a509      107  ...         9            19 -0.552096
1                1  C_ID_415bb3a509      140  ...         9            19 -0.563699
2                1  C_ID_415bb3a509      330  ...         9            14 -0.545811
3                1  C_ID_415bb3a509       -1  ...        -1             8 -0.665223
4                1  C_ID_ef55cf8d4b       -1  ...        -1            29 -0.653137

[5 rows x 15 columns]

In [ ]:
new_merchant_transactions.to_csv('New_Data_processed/new_transactions_processed.csv')

### Featurization using purchase date

Referring the EDA kernel on dates: https://www.kaggle.com/denzo123/a-closer-look-at-date-variables

In [ ]:
new_merchant_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_processed.csv', index_col=0))

Mem. usage decreased to 1582.53 Mb (52.5% reduction)


In [ ]:
new_merchant_transactions = new_merchant_transactions[['card_id', 'purchase_date', 'month_lag']]

In [ ]:
pd.to_datetime(new_merchant_transactions['purchase_date'], format='%Y-%m-%d %H:%M:%S').dt.year.value_counts()

2018    1659548
2017     303483
Name: purchase_date, dtype: int64

So our data contains transactions from two years: 2017 and 2018.

In [ ]:
new_merchant_transactions['purchase_date'] = pd.to_datetime(new_merchant_transactions['purchase_date'], format='%Y-%m-%d %H:%M:%S')

In [ ]:
# Get the day of the week for the purchase date
def get_weekday(data):
  return data.dt.dayofweek

# Return 1 if the purchase date is on a weekend
def is_weekend(day):
  if day == 5 or day == 6:
    return 1
  else:
    return 0

#Return day of purchase
def get_day(date_obj):
  return date_obj.dt.day

#Return week of the year of purchase
def get_week_of_year(date_obj):
  return date_obj.dt.weekofyear

#Return hour of purchase
def get_hour(date_obj):
  return date_obj.dt.hour

#Return month of purchase date
def get_purchase_month(data):
  return data.dt.month

#Return the month phase during the purchase
#Eg. Early in the month, mid-month or at the end of the month
def get_time_of_month(date):
  if date.day <=10:
    return "Early"
  elif date.day > 10 and date.day <= 20:
    return "Mid"
  else:
    return "End"
  

# Time of the day during purchase
#Eg. Morning, Afternoon, Evening, Night
def get_time_of_day(time):
  if time.hour >= 4 and time.hour < 12:
    return "Morning"
  elif time.hour >= 12 and time.hour < 17:
    return "Afternoon"
  elif time.hour >= 17 and time.hour < 22:
    return "Evening"
  else:
    return "Night"
    

# Returns 1 if the purchase was made on a holiday(Saturdays and sundays excluded)
# Google Search : list of holidays in brazil 2017 and 2018
def get_isholiday(date):
  holiday_list=[
            '01-01-17', '14-02-17', '28-08-17', '14-04-17', '16-04-17', '21-04-17',
            '01-05-17', '15-06-17', '07-09-17', '12-10-17', '02-11-17', '15-11-17', 
            '24-12-17', '25-12-17', '31-12-17',
            '01-01-18', '14-02-18', '28-08-18', '14-04-18', '16-04-18', '21-04-18',
            '01-05-18', '15-06-18', '07-09-18', '12-10-18', '02-11-18', '15-11-18', 
            '24-12-18', '25-12-18', '31-12-18'
  ]
  date = date.strftime(format='%d-%m-%y') 
  if date in holiday_list:
    return 1
  else:
    return 0

In [ ]:
new_merchant_transactions.head()

card_id       purchase_date  month_lag
0  C_ID_415bb3a509 2018-03-11 14:57:36          1
1  C_ID_415bb3a509 2018-03-19 18:53:37          1
2  C_ID_415bb3a509 2018-04-26 14:08:44          2
3  C_ID_415bb3a509 2018-03-07 09:43:21          1
4  C_ID_ef55cf8d4b 2018-03-22 21:07:53          1

In [ ]:
new_merchant_transactions['weekday'] = get_weekday(new_merchant_transactions['purchase_date'])
new_merchant_transactions['is_weekend'] = new_merchant_transactions['weekday'].apply(lambda day: is_weekend(day))
new_merchant_transactions['purchase_month'] = get_purchase_month(new_merchant_transactions['purchase_date'])
new_merchant_transactions['purchase_day'] = get_day(new_merchant_transactions['purchase_date'])
new_merchant_transactions['week_of_year'] = get_week_of_year(new_merchant_transactions['purchase_date'])
new_merchant_transactions['purchase_hour'] = get_hour(new_merchant_transactions['purchase_date'])
#historical_transactions['purchase_time_of_day'] = historical_transactions['purchase_date'].apply(lambda day: get_time_of_day(day))
#historical_transactions['purchase_month_time'] = historical_transactions['purchase_date'].apply(lambda date: get_time_of_month(date))
new_merchant_transactions['purchase_on_holiday'] = new_merchant_transactions['purchase_date'].apply(lambda date_obj: get_isholiday(date_obj))
new_merchant_transactions['purchase_date'] = new_merchant_transactions['purchase_date'].dt.date

In [ ]:
new_merchant_transactions.head()

card_id purchase_date  ...  purchase_hour  purchase_on_holiday
0  C_ID_415bb3a509    2018-03-11  ...             14                    0
1  C_ID_415bb3a509    2018-03-19  ...             18                    0
2  C_ID_415bb3a509    2018-04-26  ...             14                    0
3  C_ID_415bb3a509    2018-03-07  ...              9                    0
4  C_ID_ef55cf8d4b    2018-03-22  ...             21                    0

[5 rows x 10 columns]

In [ ]:
new_merchant_transactions['month_diff'] = ((datetime.date.today() - new_merchant_transactions['purchase_date']).dt.days)//30
new_merchant_transactions['month_diff'] +=new_merchant_transactions['month_lag']
del new_merchant_transactions['month_lag']
new_merchant_transactions.head()

card_id purchase_date  ...  purchase_on_holiday  month_diff
0  C_ID_415bb3a509    2018-03-11  ...                    0          32
1  C_ID_415bb3a509    2018-03-19  ...                    0          31
2  C_ID_415bb3a509    2018-04-26  ...                    0          31
3  C_ID_415bb3a509    2018-03-07  ...                    0          32
4  C_ID_ef55cf8d4b    2018-03-22  ...                    0          31

[5 rows x 10 columns]

In [ ]:
new_merchant_transactions.isna().any()

card_id                False
purchase_date          False
weekday                False
is_weekend             False
purchase_month         False
purchase_day           False
week_of_year           False
purchase_hour          False
purchase_on_holiday    False
month_diff             False
dtype: bool

Saving the dataframe due to memory constraints.

In [ ]:
new_merchant_transactions.to_csv('New_Data_processed/new_transactions_date_features.csv')

#### Aggregating:

In [ ]:
new_merchant_transactions = pd.read_csv('New_Data_processed/new_transactions_date_features.csv', index_col=0)

In [ ]:
new_merchant_transactions.head()

card_id purchase_date  ...  purchase_on_holiday  month_diff
0  C_ID_415bb3a509    2018-03-11  ...                    0          32
1  C_ID_415bb3a509    2018-03-19  ...                    0          31
2  C_ID_415bb3a509    2018-04-26  ...                    0          31
3  C_ID_415bb3a509    2018-03-07  ...                    0          32
4  C_ID_ef55cf8d4b    2018-03-22  ...                    0          31

[5 rows x 10 columns]

In [ ]:
new_merchant_transactions['purchase_date'] = pd.to_datetime(new_merchant_transactions['purchase_date'])

In [ ]:
aggregations = {
    'is_weekend': ['sum', 'mean'],
    'purchase_on_holiday': ['sum', 'mean'],
    'weekday' : ['nunique', 'sum', 'mean'],
    'purchase_hour': ['nunique', 'mean', 'min', 'max'],
    'week_of_year': ['nunique', 'mean', 'min', 'max'],
    'month_diff': ['sum', 'mean', 'min', 'max', 'var', 'skew'],
    'purchase_day': ['nunique', 'sum', 'min'],
    'purchase_date' : [np.ptp, 'min', 'max'],
    'purchase_month' : ['sum', 'mean', 'nunique']

}

aggregated_date_features = new_merchant_transactions.groupby('card_id').agg(aggregations)
aggregated_date_features.columns = ['new_transactions_'+'_'.join(col).strip() 
                           for col in aggregated_date_features.columns.values]

In [ ]:
aggregated_date_features.head()

new_transactions_is_weekend_sum  ...  new_transactions_purchase_month_nunique
card_id                                           ...                                         
C_ID_00007093c1                                0  ...                                        1
C_ID_0001238066                               12  ...                                        2
C_ID_0001506ef0                                0  ...                                        1
C_ID_0001793786                               14  ...                                        2
C_ID_000183fdda                                2  ...                                        2

[5 rows x 30 columns]

In [ ]:
aggregated_date_features['new_transactions_purchase_date_ptp'] = aggregated_date_features['new_transactions_purchase_date_ptp'].dt.days

In [ ]:
aggregated_date_features['new_transactions_purchase_date_max'].isna().any()

False

In [ ]:
d_now = datetime.datetime.today()
d_now

datetime.datetime(2020, 10, 1, 16, 1, 45, 634581)

In [ ]:
aggregated_date_features['new_transactions_purchase_date_max_diff_now'] = (d_now - aggregated_date_features['new_transactions_purchase_date_max']).dt.days

In [ ]:
aggregated_date_features['new_transactions_purchase_date_min_diff_now'] = (d_now - aggregated_date_features['new_transactions_purchase_date_min']).dt.days

In [ ]:
del aggregated_date_features['new_transactions_purchase_date_ptp']

In [ ]:
aggregated_date_features.head()

new_transactions_is_weekend_sum  ...  new_transactions_purchase_date_min_diff_now
card_id                                           ...                                             
C_ID_00007093c1                                0  ...                                          912
C_ID_0001238066                               12  ...                                          945
C_ID_0001506ef0                                0  ...                                          930
C_ID_0001793786                               14  ...                                         1051
C_ID_000183fdda                                2  ...                                          944

[5 rows x 31 columns]

In [ ]:
aggregated_date_features.columns[aggregated_date_features.isna().any()]

Index(['new_transactions_month_diff_var', 'new_transactions_month_diff_skew'], dtype='object')

In [ ]:
aggregated_date_features.to_csv('New_Data_processed/new_transactions_date_features_final.csv')

### Other Features from transactions


In [ ]:
new_merchant_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_processed.csv', index_col = 0))

Mem. usage decreased to 102.97 Mb (54.2% reduction)


In [ ]:
date_features = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_date_features.csv', index_col = 0))

Mem. usage decreased to 59.91 Mb (63.6% reduction)


In [ ]:
month_diff = date_features['month_diff']

In [ ]:
new_merchant_transactions['duration'] = new_merchant_transactions['purchase_amount']*month_diff
new_merchant_transactions['amount_month_ratio'] = new_merchant_transactions['purchase_amount']/month_diff
new_merchant_transactions['purchase_amount'] = new_merchant_transactions['purchase_amount'].apply(lambda x: min(x, 0.8))
new_merchant_transactions['installments'].replace([-1, 999], np.nan, inplace=True)
new_merchant_transactions['price'] = new_merchant_transactions['purchase_amount'] / new_merchant_transactions['installments']

In [ ]:
del month_diff,date_features

gc.collect()

0

In [ ]:
new_merchant_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_415bb3a509  ... -17.843750           -0.017426
1                1  C_ID_415bb3a509  ... -17.649414           -0.018366
2                1  C_ID_415bb3a509  ... -17.089355           -0.017783
3                1  C_ID_415bb3a509  ... -21.500000           -0.020996
4                1  C_ID_ef55cf8d4b  ... -20.449707           -0.021280

[5 rows x 17 columns]

In [ ]:
cat_aggregations = pd.DataFrame()
for col in ['category_2', 'category_3']:
  cat_aggregations[col + '_mean'] = new_merchant_transactions.groupby([col])['purchase_amount'].transform('mean')
  cat_aggregations[col + '_min'] = new_merchant_transactions.groupby([col])['purchase_amount'].transform('min')
  cat_aggregations[col + '_max'] = new_merchant_transactions.groupby([col])['purchase_amount'].transform('max')
  cat_aggregations[col + '_sum'] = new_merchant_transactions.groupby([col])['purchase_amount'].transform('sum')

In [ ]:
cat_aggregations['card_id'] = new_merchant_transactions['card_id']

In [ ]:
cat_aggregations.head()

category_2_mean  category_2_min  ...  category_3_sum          card_id
0        -0.584813        -0.74707  ...  -517323.217881  C_ID_415bb3a509
1        -0.584813        -0.74707  ...  -517323.217881  C_ID_415bb3a509
2        -0.584813        -0.74707  ...  -517323.217881  C_ID_415bb3a509
3        -0.584813        -0.74707  ...  -517323.217881  C_ID_415bb3a509
4        -0.584813        -0.74707  ...  -517323.217881  C_ID_ef55cf8d4b

[5 rows x 9 columns]

In [ ]:
new_merchant_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_415bb3a509  ... -17.843750           -0.017426
1                1  C_ID_415bb3a509  ... -17.649414           -0.018366
2                1  C_ID_415bb3a509  ... -17.089355           -0.017783
3                1  C_ID_415bb3a509  ... -21.500000           -0.020996
4                1  C_ID_ef55cf8d4b  ... -20.449707           -0.021280

[5 rows x 17 columns]

In [ ]:
cat_aggregations.to_csv('New_Data_processed/new_transactions_category_aggregations_features.csv')

In [ ]:
del cat_aggregations
gc.collect()

24

In [ ]:
new_merchant_transactions.drop(columns=['purchase_date'], axis=1, inplace=True)

In [ ]:
new_merchant_transactions.to_csv('New_Data_processed/new_transactions_processed.csv')

Saving the dataframe due to memory constraints

#### Aggregating the category features generated using transform

In [ ]:
cat_aggregations = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_category_aggregations_features.csv'))

Mem. usage decreased to 59.91 Mb (60.0% reduction)


In [ ]:
aggregated = cat_aggregations.groupby('card_id').mean()

In [ ]:
aggregated.head()

Unnamed: 0  category_2_mean  ...  category_3_max  category_3_sum
card_id                                       ...                                
C_ID_00007093c1   1793298.5        -0.582520  ...        0.799805   -517323.21875
C_ID_0001238066     38628.5        -0.584473  ...        0.799805   -445363.65625
C_ID_0001506ef0   1443419.5        -0.580078  ...        0.799805   -598762.43750
C_ID_0001793786     29154.0        -0.583008  ...        0.799805   -598762.43750
C_ID_000183fdda    827141.0        -0.580078  ...        0.799805   -347236.96875

[5 rows x 9 columns]

In [ ]:
aggregated.drop(columns=['Unnamed: 0'], axis=1, inplace=True)

In [ ]:
columns = []
for column in aggregated.columns:
  columns.append('new_' + column)

In [ ]:
columns

['new_category_2_mean',
 'new_category_2_min',
 'new_category_2_max',
 'new_category_2_sum',
 'new_category_3_mean',
 'new_category_3_min',
 'new_category_3_max',
 'new_category_3_sum']

In [ ]:
aggregated.columns = columns

In [ ]:
aggregated.columns

Index(['new_category_2_mean', 'new_category_2_min', 'new_category_2_max',
       'new_category_2_sum', 'new_category_3_mean', 'new_category_3_min',
       'new_category_3_max', 'new_category_3_sum'],
      dtype='object')

In [ ]:
aggregated.to_csv('New_Data_processed/new_transactions_category_aggregated_final.csv')

#### Numerical Features aggregation

In [ ]:
new_merchant_transactions = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_processed.csv', index_col=0))

Mem. usage decreased to 110.45 Mb (56.6% reduction)


In [ ]:
new_merchant_transactions.head()

authorized_flag          card_id  ...   duration  amount_month_ratio
0                1  C_ID_415bb3a509  ... -17.843750           -0.017426
1                1  C_ID_415bb3a509  ... -17.656250           -0.018372
2                1  C_ID_415bb3a509  ... -17.093750           -0.017776
3                1  C_ID_415bb3a509  ... -21.500000           -0.020996
4                1  C_ID_ef55cf8d4b  ... -20.453125           -0.021286

[5 rows x 16 columns]

#### Aggregating the columns based on card_id

In [ ]:
aggregations = {
    'authorized_flag' : ['sum', 'mean'],
    'category_1' : ['sum', 'mean'],
    'card_id': ['size'],
    'city_id' : ['nunique'],
    'state_id' : ['nunique'],
    'subsector_id' : ['nunique'],
    'merchant_category_id' : ['nunique'],
    'merchant_id': ['nunique'],
    'month_lag' : ['sum', 'mean', 'min', 'max', 'var'],
    'duration': ['mean', 'min', 'mean', 'max', 'var', 'skew'],
    'amount_month_ratio': ['mean', 'min', 'max', 'var', 'skew'],
    'installments' : ['sum', 'mean', 'min', 'max', 'var'],
    'purchase_amount' : ['sum', 'mean', 'min', 'max', 'var'],
    'price': ['sum', 'mean', 'min', 'max', 'var', 'skew']
}

aggregated_numerical_features = new_merchant_transactions.groupby('card_id').agg(aggregations)
aggregated_numerical_features.columns = ['new_transactions_'+'_'.join(col).strip() 
                           for col in aggregated_numerical_features.columns.values]

In [ ]:
aggregated_numerical_features.head()

new_transactions_authorized_flag_sum  ...  new_transactions_price_skew
card_id                                                ...                             
C_ID_00007093c1                                     2  ...                          NaN
C_ID_0001238066                                    26  ...                          NaN
C_ID_0001506ef0                                     2  ...                          NaN
C_ID_0001793786                                    31  ...                          NaN
C_ID_000183fdda                                    11  ...                          NaN

[5 rows x 42 columns]

In [ ]:
aggregated_numerical_features.to_csv('New_Data_processed/new_transactions_other_features_final.csv')

# Merchants File

#### Joining with the card_ids

In [ ]:
merchants = reduce_mem_usage(pd.read_csv('Data/merchants.csv'))
merchants.drop(columns=['merchant_category_id','subsector_id', 'city_id', 'state_id'], axis = 1, inplace=True)

Mem. usage decreased to 30.32 Mb (46.0% reduction)


In [ ]:
merged_merchants_with_cards = reduce_mem_usage(pd.read_csv('Data_processed/all_transactions.csv', index_col = 0))
merged_merchants_with_cards = merged_merchants_with_cards[['card_id', 'merchant_id']]

In [ ]:
merged_merchants_with_cards = pd.merge(merged_merchants_with_cards, merchants, on='merchant_id', how='left')

In [ ]:
merged_merchants_with_cards.head()

card_id      merchant_id  ...  category_4  category_2
0  C_ID_4e6213e9bc  M_ID_e020e9b302  ...           Y         1.0
1  C_ID_4e6213e9bc  M_ID_86ec983688  ...           Y         1.0
2  C_ID_4e6213e9bc  M_ID_979ed661fc  ...           Y         1.0
3  C_ID_4e6213e9bc  M_ID_e6d5ae8ea6  ...           Y         1.0
4  C_ID_4e6213e9bc  M_ID_e020e9b302  ...           Y         1.0

[5 rows x 19 columns]

In [ ]:
merged_merchants_categorical_features = merged_merchants_with_cards[['card_id', 'merchant_group_id',
                                                                   'category_1', 'most_recent_sales_range',
                                                                   'most_recent_purchases_range', 'active_months_lag3',
                                                                   'active_months_lag6', 'active_months_lag12',
                                                                   'category_4', 'category_2',
                                                                   ]]
                                                                  
merged_merchants_numerical_features = merged_merchants_with_cards[['card_id', 'numerical_1', 'numerical_2',
                                                                   'avg_sales_lag3', 'avg_purchases_lag3',
                                                                   'avg_sales_lag6', 'avg_purchases_lag6',
                                                                   'avg_sales_lag12', 'avg_purchases_lag12',]]                                                                  

In [ ]:
merged_merchants_categorical_features.to_csv('Data_processed/merged_merchants_categorical_features.csv')
merged_merchants_numerical_features.to_csv('Data_processed/merged_merchants_numerical_features.csv')

#### Categorical Features

In [ ]:
merged_merchants_categorical_features = reduce_mem_usage(pd.read_csv('Data_processed/merged_merchants_categorical_features.csv', index_col=0))

Mem. usage decreased to 1849.86 Mb (31.8% reduction)


In [ ]:
merged_merchants_categorical_features.isna().any()

card_id                        False
merchant_group_id               True
category_1                      True
most_recent_sales_range         True
most_recent_purchases_range     True
active_months_lag3              True
active_months_lag6              True
active_months_lag12             True
category_4                      True
category_2                      True
dtype: bool

In [ ]:
merged_merchants_categorical_features['merchant_group_id'].fillna(merged_merchants_categorical_features['merchant_group_id'].mode()[0], inplace=True)
merged_merchants_categorical_features['category_1'].fillna(merged_merchants_categorical_features['category_1'].mode()[0], inplace=True)
merged_merchants_categorical_features['most_recent_sales_range'].fillna(merged_merchants_categorical_features['most_recent_sales_range'].mode()[0], inplace=True)
merged_merchants_categorical_features['most_recent_purchases_range'].fillna(merged_merchants_categorical_features['most_recent_purchases_range'].mode()[0], inplace=True)
merged_merchants_categorical_features['active_months_lag3'].fillna(merged_merchants_categorical_features['active_months_lag3'].mode()[0], inplace=True)
merged_merchants_categorical_features['active_months_lag6'].fillna(merged_merchants_categorical_features['active_months_lag6'].mode()[0], inplace=True)
merged_merchants_categorical_features['active_months_lag12'].fillna(merged_merchants_categorical_features['active_months_lag12'].mode()[0], inplace=True)
merged_merchants_categorical_features['category_4'].fillna(merged_merchants_categorical_features['category_4'].mode()[0], inplace=True)
merged_merchants_categorical_features['category_2'].fillna(merged_merchants_categorical_features['category_2'].mode()[0], inplace=True)

In [ ]:
merged_merchants_categorical_features.isna().any()

card_id                        False
merchant_group_id              False
category_1                     False
most_recent_sales_range        False
most_recent_purchases_range    False
active_months_lag3             False
active_months_lag6             False
active_months_lag12            False
category_4                     False
category_2                     False
dtype: bool

In [ ]:
merged_merchants_categorical_features.head()

card_id  merchant_group_id  ... category_4 category_2
0  C_ID_4e6213e9bc               35.0  ...          Y        1.0
1  C_ID_4e6213e9bc             2084.0  ...          Y        1.0
2  C_ID_4e6213e9bc            27369.0  ...          Y        1.0
3  C_ID_4e6213e9bc            24104.0  ...          Y        1.0
4  C_ID_4e6213e9bc               35.0  ...          Y        1.0

[5 rows x 10 columns]

In [ ]:
merged_merchants_categorical_features['category_1'] = merged_merchants_categorical_features['category_1'].map({
    'Y': 1,
    'N': 0
})
merged_merchants_categorical_features['category_4'] = merged_merchants_categorical_features['category_4'].map({
    'Y': 1,
    'N': 0
})

In [ ]:
merged_merchants_categorical_features = pd.get_dummies(merged_merchants_categorical_features,columns=[
            'most_recent_sales_range','most_recent_purchases_range', 'active_months_lag3', 'active_months_lag6',
            'active_months_lag12', 'category_2'
])

In [ ]:
merged_merchants_categorical_features.head()

card_id  merchant_group_id  ...  category_2_4.0  category_2_5.0
0  C_ID_4e6213e9bc               35.0  ...               0               0
1  C_ID_4e6213e9bc             2084.0  ...               0               0
2  C_ID_4e6213e9bc            27369.0  ...               0               0
3  C_ID_4e6213e9bc            24104.0  ...               0               0
4  C_ID_4e6213e9bc               35.0  ...               0               0

[5 rows x 40 columns]

In [ ]:
aggregations = {
    'merchant_group_id' : ['nunique'],
    'category_1' : ['sum', 'mean', 'std'],
    'category_4' :['sum', 'mean', 'std'],
    'category_2_1.0' :['sum', 'mean', 'std'],
    'category_2_2.0' :['sum', 'mean', 'std'],
    'category_2_3.0' :['sum', 'mean', 'std'],
    'category_2_4.0' :['sum', 'mean', 'std'],
    'category_2_5.0' :['sum', 'mean', 'std'],
    'most_recent_sales_range_A' :['sum', 'mean', 'std'],
    'most_recent_sales_range_B' :['sum', 'mean', 'std'],
    'most_recent_sales_range_C' :['sum', 'mean', 'std'],
    'most_recent_sales_range_D' :['sum', 'mean', 'std'],
    'most_recent_sales_range_E' :['sum', 'mean', 'std'],
    'most_recent_purchases_range_A' :['sum', 'mean', 'std'],
    'most_recent_purchases_range_B' :['sum', 'mean', 'std'],
    'most_recent_purchases_range_C' :['sum', 'mean', 'std'],
    'most_recent_purchases_range_D' :['sum', 'mean', 'std'],
    'most_recent_purchases_range_E' :['sum', 'mean', 'std'],
    'active_months_lag3_1.0' :['sum', 'mean', 'std'],
    'active_months_lag3_2.0' :['sum', 'mean', 'std'],
    'active_months_lag3_3.0' :['sum', 'mean', 'std'],
    'active_months_lag6_1.0' :['sum', 'mean', 'std'],
    'active_months_lag6_2.0' :['sum', 'mean', 'std'],
    'active_months_lag6_3.0' :['sum', 'mean', 'std'],
    'active_months_lag6_4.0' :['sum', 'mean', 'std'],
    'active_months_lag6_5.0' :['sum', 'mean', 'std'],
    'active_months_lag6_6.0' :['sum', 'mean', 'std'],
    'active_months_lag12_1.0' :['sum', 'mean', 'std'],
    'active_months_lag12_2.0' :['sum', 'mean', 'std'],
    'active_months_lag12_3.0' :['sum', 'mean', 'std'],
    'active_months_lag12_4.0' :['sum', 'mean', 'std'],
    'active_months_lag12_5.0' :['sum', 'mean', 'std'],
    'active_months_lag12_6.0' :['sum', 'mean', 'std'],
    'active_months_lag12_7.0' :['sum', 'mean', 'std'],
    'active_months_lag12_8.0' :['sum', 'mean', 'std'],
    'active_months_lag12_9.0' :['sum', 'mean', 'std'],
    'active_months_lag12_10.0' :['sum', 'mean', 'std'],
    'active_months_lag12_11.0' :['sum', 'mean', 'std'],
    'active_months_lag12_12.0' :['sum', 'mean', 'std']
}

merchant_features = merged_merchants_categorical_features.groupby('card_id').agg(aggregations)
merchant_features.columns = ['merchant_'+'_'.join(col).strip() 
                           for col in merchant_features.columns.values]

In [ ]:
merchant_features.head()

merchant_merchant_group_id_nunique  ...  merchant_active_months_lag12_12.0_std
card_id                                              ...                                       
C_ID_00007093c1                                  15  ...                               0.178951
C_ID_0001238066                                  66  ...                               0.140469
C_ID_0001506ef0                                  19  ...                               0.167802
C_ID_0001793786                                 104  ...                               0.204335
C_ID_000183fdda                                  64  ...                               0.190580

[5 rows x 115 columns]

In [ ]:
merchant_features.to_csv('New_Data_processed/merchant_categorical_features_final.csv')

#### Numerical Features

In [ ]:
merged_merchants_numerical_features = reduce_mem_usage(pd.read_csv('Data_processed/merged_merchants_numerical_features.csv', index_col=0))

Mem. usage decreased to 1726.54 Mb (30.0% reduction)


In [ ]:
merged_merchants_numerical_features.isna().any()

card_id                False
numerical_1             True
numerical_2             True
avg_sales_lag3          True
avg_purchases_lag3      True
avg_sales_lag6          True
avg_purchases_lag6      True
avg_sales_lag12         True
avg_purchases_lag12     True
dtype: bool

In [ ]:
merged_merchants_numerical_features['numerical_1'].fillna(merged_merchants_numerical_features['numerical_1'].mode()[0], inplace=True)
merged_merchants_numerical_features['numerical_2'].fillna(merged_merchants_numerical_features['numerical_2'].mode()[0], inplace=True)
merged_merchants_numerical_features['avg_sales_lag3'].fillna(merged_merchants_numerical_features['avg_sales_lag3'].mean(), inplace=True)
merged_merchants_numerical_features['avg_purchases_lag3'].fillna(merged_merchants_numerical_features['avg_purchases_lag3'].mean(), inplace=True)
merged_merchants_numerical_features['avg_sales_lag6'].fillna(merged_merchants_numerical_features['avg_sales_lag6'].mean(), inplace=True)
merged_merchants_numerical_features['avg_purchases_lag6'].fillna(merged_merchants_numerical_features['avg_purchases_lag6'].mean(), inplace=True)
merged_merchants_numerical_features['avg_sales_lag12'].fillna(merged_merchants_numerical_features['avg_sales_lag12'].mean(), inplace=True)
merged_merchants_numerical_features['avg_purchases_lag12'].fillna(merged_merchants_numerical_features['avg_purchases_lag12'].mean(), inplace=True)

In [ ]:
merged_merchants_numerical_features.isna().any()

card_id                False
numerical_1            False
numerical_2            False
avg_sales_lag3         False
avg_purchases_lag3     False
avg_sales_lag6         False
avg_purchases_lag6     False
avg_sales_lag12        False
avg_purchases_lag12    False
dtype: bool

In [ ]:
merged_merchants_numerical_features.head()

card_id  numerical_1  ...  avg_sales_lag12  avg_purchases_lag12
0  C_ID_4e6213e9bc    23.468750  ...             1.19             1.156844
1  C_ID_4e6213e9bc     1.092773  ...             1.05             1.062087
2  C_ID_4e6213e9bc     0.021851  ...             0.97             0.956668
3  C_ID_4e6213e9bc    -0.057465  ...             0.86             0.864394
4  C_ID_4e6213e9bc    23.468750  ...             1.19             1.156844

[5 rows x 9 columns]

In [ ]:
aggregations = {
    'numerical_1': ['sum', 'mean', 'min', 'max', 'std'],
    'numerical_2': ['sum', 'mean', 'min', 'max', 'std'],
    'avg_sales_lag3': ['sum', 'mean', 'std'],
    'avg_purchases_lag3': ['sum', 'mean', 'std'],
    'avg_sales_lag6': ['sum', 'mean', 'std'],
    'avg_purchases_lag6': ['sum', 'mean', 'std'],
    'avg_sales_lag12': ['sum', 'mean', 'std'],
    'avg_purchases_lag12': ['sum', 'mean', 'std'],
}

merchant_features = merged_merchants_numerical_features.groupby('card_id').agg(aggregations)
merchant_features.columns = ['merchant_'+'_'.join(col).strip() 
                           for col in merchant_features.columns.values]

In [ ]:
merchant_features.head()

merchant_numerical_1_sum  ...  merchant_avg_purchases_lag12_std
card_id                                    ...                                  
C_ID_00007093c1                  270.0000  ...                          1.333689
C_ID_0001238066                 1324.0000  ...                               NaN
C_ID_0001506ef0                   65.4375  ...                               NaN
C_ID_0001793786                  287.2500  ...                         37.842564
C_ID_000183fdda                   32.5000  ...                         61.670996

[5 rows x 28 columns]

In [ ]:
merchant_features.to_csv('New_Data_processed/merchant_numerical_features_final.csv')

# Merging all features together

In [ ]:
train = reduce_mem_usage(pd.read_csv('New_Data_processed/processed_train.csv', index_col=0))
test = reduce_mem_usage(pd.read_csv('New_Data_processed/processed_test.csv', index_col=0))

Mem. usage decreased to 13.48 Mb (67.6% reduction)
Mem. usage decreased to  8.84 Mb (63.9% reduction)


In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201917 entries, 0 to 201916
Data columns (total 26 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   first_active_month                  201917 non-null  object 
 1   card_id                             201917 non-null  object 
 2   feature_1                           201917 non-null  float16
 3   feature_2                           201917 non-null  float16
 4   feature_3                           201917 non-null  float16
 5   target                              201917 non-null  float16
 6   quarter_first_active_month          201917 non-null  int8   
 7   first_active_month_diff_from_today  201917 non-null  int16  
 8   rare_datapoints                     201917 non-null  int8   
 9   cat_time_feature1                   201917 non-null  float16
 10  cat_time_feature2                   201917 non-null  float16
 11  cat_time_feature3         

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 123623 entries, 0 to 123622
Data columns (total 25 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   first_active_month                  123623 non-null  object 
 1   card_id                             123623 non-null  object 
 2   feature_1                           123623 non-null  float16
 3   feature_2                           123623 non-null  float16
 4   feature_3                           123623 non-null  float16
 5   quarter_first_active_month          123623 non-null  int8   
 6   first_active_month_diff_from_today  123623 non-null  int16  
 7   target                              0 non-null       float64
 8   cat_time_feature1                   123623 non-null  float16
 9   cat_time_feature2                   123623 non-null  float16
 10  cat_time_feature3                   123623 non-null  float16
 11  cat_time_ratio1           

### Merging Transaction Features

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/historical_transactions_date_features_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to 23.28 Mb (70.7% reduction)


first_active_month  ... transactions_purchase_date_min_diff_now
0         2017-06-01  ...                                    1190
1         2017-01-01  ...                                    1362
2         2016-08-01  ...                                    1357
3         2017-09-01  ...                                    1099
4         2017-11-01  ...                                    1052

[5 rows x 57 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... transactions_purchase_date_min_diff_now
0         2017-04-01  ...                                    1274
1         2017-01-01  ...                                    1355
2         2017-08-01  ...                                    1131
3         2017-12-01  ...                                    1030
4         2015-12-01  ...                                    1365

[5 rows x 56 columns]

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_date_features_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to 18.81 Mb (73.4% reduction)


first_active_month  ... new_transactions_purchase_date_min_diff_now
0         2017-06-01  ...                                       941.0
1         2017-01-01  ...                                       973.0
2         2016-08-01  ...                                       887.0
3         2017-09-01  ...                                       939.0
4         2017-11-01  ...                                       944.0

[5 rows x 88 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... new_transactions_purchase_date_min_diff_now
0         2017-04-01  ...                                       971.0
1         2017-01-01  ...                                       943.0
2         2017-08-01  ...                                       945.0
3         2017-12-01  ...                                       942.0
4         2015-12-01  ...                                       942.0

[5 rows x 87 columns]

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/transactions_category_aggregated_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to  8.69 Mb (61.1% reduction)


first_active_month          card_id  ...  category_3_max  category_3_sum
0         2017-06-01  C_ID_92a2005557  ...        0.799805      -9929442.0
1         2017-01-01  C_ID_3d0044924f  ...        0.799805      -5854655.0
2         2016-08-01  C_ID_d639edf6cd  ...        0.799805      -9971130.0
3         2017-09-01  C_ID_186d6a6901  ...        0.799805      -6703284.0
4         2017-11-01  C_ID_cdbd2c0db2  ...        0.799805      -6897541.5

[5 rows x 96 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month          card_id  ...  category_3_max  category_3_sum
0         2017-04-01  C_ID_0ab67a22ab  ...        0.799805      -5024630.5
1         2017-01-01  C_ID_130fd0cbdd  ...        0.799805      -7084146.5
2         2017-08-01  C_ID_b709037bc5  ...        0.799805      -3100319.0
3         2017-12-01  C_ID_d27d835a9f  ...        0.799805      -5608475.0
4         2015-12-01  C_ID_2b5e3df5c2  ...        0.799805      -6884309.5

[5 rows x 95 columns]

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_category_aggregated_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to  7.74 Mb (61.1% reduction)


first_active_month          card_id  ...  new_category_3_max  new_category_3_sum
0         2017-06-01  C_ID_92a2005557  ...            0.799805       -598762.43750
1         2017-01-01  C_ID_3d0044924f  ...            0.799805       -517323.21875
2         2016-08-01  C_ID_d639edf6cd  ...            0.799805       -598762.43750
3         2017-09-01  C_ID_186d6a6901  ...            0.799805       -528957.37500
4         2017-11-01  C_ID_cdbd2c0db2  ...            0.799805       -506027.15625

[5 rows x 104 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month          card_id  ...  new_category_3_max  new_category_3_sum
0         2017-04-01  C_ID_0ab67a22ab  ...            0.799805       -354624.21875
1         2017-01-01  C_ID_130fd0cbdd  ...            0.799805       -463090.21875
2         2017-08-01  C_ID_b709037bc5  ...            0.799805       -273274.71875
3         2017-12-01  C_ID_d27d835a9f  ...            0.799805       -273274.71875
4         2015-12-01  C_ID_2b5e3df5c2  ...            0.799805       -449546.90625

[5 rows x 103 columns]

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/transactions_other_features_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to 32.29 Mb (69.8% reduction)


first_active_month  ... transactions_price_skew
0         2017-06-01  ...                     inf
1         2017-01-01  ...                    -inf
2         2016-08-01  ...                     inf
3         2017-09-01  ...                    -inf
4         2017-11-01  ...                 2.40625

[5 rows x 146 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... transactions_price_skew
0         2017-04-01  ...                0.891602
1         2017-01-01  ...                3.214844
2         2017-08-01  ...                     inf
3         2017-12-01  ...                4.902344
4         2015-12-01  ...                1.742188

[5 rows x 145 columns]

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/new_transactions_other_features_final.csv', index_col=0))
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

Mem. usage decreased to 30.70 Mb (67.7% reduction)


first_active_month  ... new_transactions_price_skew
0         2017-06-01  ...                         NaN
1         2017-01-01  ...                    1.203125
2         2016-08-01  ...                         NaN
3         2017-09-01  ...                         NaN
4         2017-11-01  ...                         NaN

[5 rows x 188 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... new_transactions_price_skew
0         2017-04-01  ...                    1.686523
1         2017-01-01  ...                    2.683594
2         2017-08-01  ...                         NaN
3         2017-12-01  ...                    0.031342
4         2015-12-01  ...                         NaN

[5 rows x 187 columns]

# Adding some additional features:
<br>
Ref: https://www.kaggle.com/chauhuynh/my-first-kernel-3-699

In [ ]:
train['transactions_purchase_date_max'] = pd.to_datetime(train['transactions_purchase_date_max'])
test['transactions_purchase_date_max'] = pd.to_datetime(test['transactions_purchase_date_max'])
train['new_transactions_purchase_date_max'] = pd.to_datetime(train['new_transactions_purchase_date_max'])
test['new_transactions_purchase_date_max'] = pd.to_datetime(test['new_transactions_purchase_date_max'])

train['transactions_purchase_date_min'] = pd.to_datetime(train['transactions_purchase_date_min'])
test['transactions_purchase_date_min'] = pd.to_datetime(test['transactions_purchase_date_min'])
train['new_transactions_purchase_date_min'] = pd.to_datetime(train['new_transactions_purchase_date_min'])
test['new_transactions_purchase_date_min'] = pd.to_datetime(test['new_transactions_purchase_date_min'])

train['first_active_month'] = pd.to_datetime(train['first_active_month'])
test['first_active_month'] = pd.to_datetime(test['first_active_month'])

In [ ]:
train['purchase_date_diff'] = (train['transactions_purchase_date_max'] - train['transactions_purchase_date_min']).dt.days
test['purchase_date_diff'] = (test['transactions_purchase_date_max'] - test['transactions_purchase_date_min']).dt.days
train['new_purchase_date_diff'] = (train['new_transactions_purchase_date_max'] - train['new_transactions_purchase_date_min']).dt.days
test['new_purchase_date_diff'] = (test['new_transactions_purchase_date_max'] - test['new_transactions_purchase_date_min']).dt.days

train['purchase_date_average'] = (train['purchase_date_diff'])/train['transactions_card_id_size']
test['purchase_date_average'] = (test['purchase_date_diff'])/test['transactions_card_id_size']
train['new_purchase_date_average'] = (train['new_purchase_date_diff'])/train['new_transactions_card_id_size']
test['new_purchase_date_average'] = (test['new_purchase_date_diff'])/test['new_transactions_card_id_size']

train['purchase_date_diff_now'] = (datetime.datetime.today() - train['transactions_purchase_date_max']).dt.days
test['purchase_date_diff_now'] = (datetime.datetime.today() - test['transactions_purchase_date_max']).dt.days
train['new_purchase_date_diff_now'] = (datetime.datetime.today() - train['new_transactions_purchase_date_max']).dt.days
test['new_purchase_date_diff_now'] = (datetime.datetime.today() - test['new_transactions_purchase_date_max']).dt.days

train['purchase_date_diff_now_min'] = (datetime.datetime.today() - train['transactions_purchase_date_min']).dt.days
test['purchase_date_diff_now_min'] = (datetime.datetime.today() - test['transactions_purchase_date_min']).dt.days
train['new_purchase_date_diff_now_min'] = (datetime.datetime.today() - train['new_transactions_purchase_date_min']).dt.days
test['new_purchase_date_diff_now_min'] = (datetime.datetime.today() - test['new_transactions_purchase_date_min']).dt.days

train['first_buy'] = (train['transactions_purchase_date_min'] - train['first_active_month']).dt.days
test['first_buy'] = (test['transactions_purchase_date_min'] - test['first_active_month']).dt.days
train['new_first_buy'] = (train['new_transactions_purchase_date_min'] - train['first_active_month']).dt.days
test['new_first_buy'] = (test['new_transactions_purchase_date_min'] - test['first_active_month']).dt.days

train['last_buy'] = (train['transactions_purchase_date_max'] - train['first_active_month']).dt.days
test['last_buy'] = (test['transactions_purchase_date_max'] - test['first_active_month']).dt.days
train['new_last_buy'] = (train['new_transactions_purchase_date_max'] - train['first_active_month']).dt.days
test['new_last_buy'] = (test['new_transactions_purchase_date_max'] - test['first_active_month']).dt.days

In [ ]:
train['transactions_purchase_date_max'] = train['transactions_purchase_date_max'].astype(np.int64) * 1e-9
train['new_transactions_purchase_date_max'] = train['new_transactions_purchase_date_max'].astype(np.int64) * 1e-9
train['transactions_purchase_date_min'] = train['transactions_purchase_date_min'].astype(np.int64) * 1e-9
train['new_transactions_purchase_date_min'] = train['new_transactions_purchase_date_min'].astype(np.int64) * 1e-9

test['transactions_purchase_date_max'] = test['transactions_purchase_date_max'].astype(np.int64) * 1e-9
test['new_transactions_purchase_date_max'] = test['new_transactions_purchase_date_max'].astype(np.int64) * 1e-9
test['transactions_purchase_date_min'] = test['transactions_purchase_date_min'].astype(np.int64) * 1e-9
test['new_transactions_purchase_date_min'] = test['new_transactions_purchase_date_min'].astype(np.int64) * 1e-9

In [ ]:
train['card_id_total'] = train['new_transactions_card_id_size'] + train['transactions_card_id_size'] 
test['card_id_total'] = test['new_transactions_card_id_size'] + test['transactions_card_id_size'] 
train['card_id_ratio'] = train['new_transactions_card_id_size'] / train['transactions_card_id_size']
test['card_id_ratio'] = test['new_transactions_card_id_size'] / test['transactions_card_id_size']

train['purchase_amount_total'] = train['new_transactions_purchase_amount_sum'] + train['transactions_purchase_amount_sum']
train['purchase_amount_mean'] = train['new_transactions_purchase_amount_mean'] + train['transactions_purchase_amount_mean']
train['purchase_amount_max'] = train['new_transactions_purchase_amount_max'] + train['transactions_purchase_amount_max']
train['purchase_amount_min'] = train['new_transactions_purchase_amount_min'] + train['transactions_purchase_amount_min']
train['purchase_amount_ratio'] = train['new_transactions_purchase_amount_sum'] / train['transactions_purchase_amount_sum']
test['purchase_amount_total'] = test['new_transactions_purchase_amount_sum'] + test['transactions_purchase_amount_sum']
test['purchase_amount_mean'] = test['new_transactions_purchase_amount_mean'] + test['transactions_purchase_amount_mean']
test['purchase_amount_max'] = test['new_transactions_purchase_amount_max'] + test['transactions_purchase_amount_max']
test['purchase_amount_min'] = test['new_transactions_purchase_amount_min'] + test['transactions_purchase_amount_min']
test['purchase_amount_ratio'] = test['new_transactions_purchase_amount_sum'] / test['transactions_purchase_amount_sum']

train['month_diff_mean'] = train['new_transactions_month_diff_mean'] + train['transactions_month_diff_mean']
test['month_diff_mean'] = test['new_transactions_month_diff_mean'] + test['transactions_month_diff_mean']
train['month_diff_ratio'] = train['new_transactions_month_diff_mean'] / train['transactions_month_diff_mean']
test['month_diff_ratio'] = test['new_transactions_month_diff_mean'] / test['transactions_month_diff_mean']
train['month_lag_mean'] = train['new_transactions_month_lag_mean'] + train['transactions_month_lag_mean']
test['month_lag_mean'] = test['new_transactions_month_lag_mean'] + test['transactions_month_lag_mean']
train['month_lag_max'] = train['new_transactions_month_lag_max'] + train['transactions_month_lag_max']
test['month_lag_max'] = test['new_transactions_month_lag_max'] + test['transactions_month_lag_max']
train['month_lag_min'] = train['new_transactions_month_lag_min'] + train['transactions_month_lag_min']
test['month_lag_min'] = test['new_transactions_month_lag_min'] + test['transactions_month_lag_min']

train['category_1_mean'] = train['new_transactions_category_1_mean'] + train['transactions_category_1_mean']
test['category_1_mean'] = test['new_transactions_category_1_mean'] + test['transactions_category_1_mean']
train['category_1_sum'] = train['new_transactions_category_1_sum'] + train['transactions_category_1_sum']
test['category_1_sum'] = test['new_transactions_category_1_sum'] + test['transactions_category_1_sum']

train['installments_mean'] = train['new_transactions_installments_mean'] + train['transactions_installments_mean']
test['installments_mean'] = test['new_transactions_installments_mean'] + test['transactions_installments_mean']
train['installments_total'] = train['new_transactions_installments_sum'] + train['transactions_installments_sum']
test['installments_total'] = test['new_transactions_installments_sum'] + test['transactions_installments_sum']
train['installments_ratio'] = train['new_transactions_installments_sum'] / train['transactions_installments_sum']
test['installments_ratio'] = test['new_transactions_installments_sum'] / test['transactions_installments_sum']
train['installments_max'] = train['new_transactions_installments_max'] + train['transactions_installments_max']
test['installments_max'] = test['new_transactions_installments_max'] + test['transactions_installments_max']
train['installments_min'] = train['new_transactions_installments_min'] + train['transactions_installments_min']
test['installments_min'] = train['new_transactions_installments_min'] + train['transactions_installments_min']

train['duration_mean'] = train['new_transactions_duration_mean'] + train['transactions_duration_mean']
test['duration_mean'] = test['new_transactions_duration_mean'] + test['transactions_duration_mean']
train['duration_max'] = train['new_transactions_duration_max'] + train['transactions_duration_max']
test['duration_max'] = test['new_transactions_duration_max'] + test['transactions_duration_max']
train['duration_min'] = train['new_transactions_duration_min'] + train['transactions_duration_min']
test['duration_min'] = test['new_transactions_duration_min'] + test['transactions_duration_min']

train['amount_month_ratio_mean'] = train['new_transactions_amount_month_ratio_mean'] + train['transactions_amount_month_ratio_mean']
test['amount_month_ratio_mean'] = test['new_transactions_amount_month_ratio_mean'] + test['transactions_amount_month_ratio_mean']
train['amount_month_ratio_min'] = train['new_transactions_amount_month_ratio_min'] + train['transactions_amount_month_ratio_min']
test['amount_month_ratio_min'] = test['new_transactions_amount_month_ratio_min'] + test['transactions_amount_month_ratio_min']
train['amount_month_ratio_max'] = train['new_transactions_amount_month_ratio_max'] + train['transactions_amount_month_ratio_max']
test['amount_month_ratio_max'] = test['new_transactions_amount_month_ratio_max'] + test['transactions_amount_month_ratio_max']

train['CLV'] = train['transactions_card_id_size'] * train['transactions_purchase_amount_sum'] / train['transactions_month_diff_mean']
test['CLV'] = test['transactions_card_id_size'] * test['transactions_purchase_amount_sum'] / test['transactions_month_diff_mean']
train['new_CLV'] = train['new_transactions_card_id_size'] * train['new_transactions_purchase_amount_sum'] / train['new_transactions_month_diff_mean']
test['new_CLV'] = test['new_transactions_card_id_size'] * test['new_transactions_purchase_amount_sum'] / test['new_transactions_month_diff_mean']

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201917 entries, 0 to 201916
Columns: 227 entries, first_active_month to new_CLV
dtypes: datetime64[ns](1), float16(128), float32(5), float64(54), int16(7), int32(3), int64(7), int8(21), object(1)
memory usage: 170.8+ MB


In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 123623 entries, 0 to 123622
Columns: 226 entries, first_active_month to new_CLV
dtypes: datetime64[ns](1), float16(127), float32(5), float64(57), int16(7), int32(3), int64(5), int8(20), object(1)
memory usage: 99.0+ MB


### Merging Merchant Features

#### Merging Numerical Features

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/merchant_numerical_features_final.csv', index_col=0))

Mem. usage decreased to 36.63 Mb (49.1% reduction)


In [ ]:
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

first_active_month  ... merchant_avg_purchases_lag12_std
0         2017-06-01  ...                              NaN
1         2017-01-01  ...                        40.625000
2         2016-08-01  ...                         2.261719
3         2017-09-01  ...                         4.460938
4         2017-11-01  ...                              NaN

[5 rows x 165 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... merchant_avg_purchases_lag12_std
0         2017-04-01  ...                         3.099609
1         2017-01-01  ...                         1.655273
2         2017-08-01  ...                         0.266846
3         2017-12-01  ...                         0.136475
4         2015-12-01  ...                         9.312500

[5 rows x 165 columns]

#### Merging Categorical Features

In [ ]:
to_join = reduce_mem_usage(pd.read_csv('New_Data_processed/merchant_categorical_features_final.csv', index_col=0))

Mem. usage decreased to 69.85 Mb (75.8% reduction)


In [ ]:
train = pd.merge(train, to_join, on='card_id', how='left')
train.head()

first_active_month  ... merchant_active_months_lag12_12.0_std
0         2017-06-01  ...                              0.164673
1         2017-01-01  ...                              0.177124
2         2016-08-01  ...                              0.208374
3         2017-09-01  ...                              0.354736
4         2017-11-01  ...                              0.251465

[5 rows x 280 columns]

In [ ]:
test = pd.merge(test, to_join, on='card_id', how='left')
test.head()

first_active_month  ... merchant_active_months_lag12_12.0_std
0         2017-04-01  ...                              0.198608
1         2017-01-01  ...                              0.149902
2         2017-08-01  ...                              0.000000
3         2017-12-01  ...                              0.000000
4         2015-12-01  ...                              0.278564

[5 rows x 280 columns]

Check for NA and imputating the NA values if needed

In [ ]:
na_columns = train.columns[train.isna().any()]

In [ ]:
for column in na_columns:
  train[column].fillna(train[column].mode()[0], inplace=True)
  test[column].fillna(test[column].mode()[0], inplace=True)

In [ ]:
train.head()

first_active_month          card_id  ...          CLV    new_CLV
0         2017-06-01  C_ID_92a2005557  ... -1382.193604  -9.664105
1         2017-01-01  C_ID_3d0044924f  ... -2326.133057  -0.804087
2         2016-08-01  C_ID_d639edf6cd  ...   -40.040398  -0.022587
3         2017-09-01  C_ID_186d6a6901  ...  -122.954704  -1.032162
4         2017-11-01  C_ID_cdbd2c0db2  ...  -319.146973 -22.711529

[5 rows x 227 columns]

## Check for Nan and impute if needed

In [ ]:
train.columns[train.isna().any()]

Index([], dtype='object')

In [ ]:
test.columns[test.isna().any()]

Index(['target', 'new_purchase_date_diff_now', 'new_last_buy'], dtype='object')

In [ ]:
train.columns[train.isna().any()]

Index([], dtype='object')

In [ ]:
test.columns[test.isna().any()]

Index(['target'], dtype='object')

In [ ]:
train.drop(columns=['first_active_month'], inplace=True)
test.drop(columns=['first_active_month'], inplace=True)

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201917 entries, 0 to 201916
Columns: 226 entries, card_id to new_CLV
dtypes: float16(128), float32(5), float64(54), int16(7), int32(3), int64(7), int8(21), object(1)
memory usage: 169.2+ MB


Saving two sets of files. One with the merchant features and one without it. 

In [ ]:
train.to_csv('New_Data_processed/train_features_without_merchant.csv')
test.to_csv('New_Data_processed/test_features_without_merchant.csv')

In [ ]:
train.to_csv('New_Data_processed/train_features_generated.csv')
test.to_csv('New_Data_processed/test_features_generated.csv')